aten/src/ATen/OptionsGuard.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <ATen/DefaultTensorOptions.h>
+#include <ATen/TensorOptions.h>
+
+namespace at {
+
+/// RAII guard that stores the current default options upon construction, sets
+/// the current default options to the ones given to its constructor, and
+/// finally resets the options back to the original ones in the destructor.
+struct OptionsGuard {
```

Comment:
RAII guards should be reentrant, right? If your default options is global state, this is not reentrant.